<a href="https://colab.research.google.com/github/Lital-Ben-Baruch/Data_Visualization/blob/main/Part_A_knowing_your_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part A - knowing your data

### Step 1. Import the necessary libraries

In [24]:
import pandas as pd

### Step 2. Import the dataset from drive

In [25]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Step 3. Assign it to a variable called df_occupation.

In [26]:
df_occupation = pd.read_csv('/content/drive/My Drive/Summaries _ML_AI_CV/data_sets/pandas/occupation.csv', sep = ',')

### Step 4. See the first 25 entries

In [27]:
df_occupation.head(25)

user_id  age gender     occupation zip_code
0         1   24      M     technician    85711
1         2   53      F          other    94043
2         3   23      M         writer    32067
3         4   24      M     technician    43537
4         5   33      F          other    15213
5         6   42      M      executive    98101
6         7   57      M  administrator    91344
7         8   36      M  administrator    05201
8         9   29      M        student    01002
9        10   53      M         lawyer    90703
10       11   39      F          other    30329
11       12   28      F          other    06405
12       13   47      M       educator    29206
13       14   45      M      scientist    55106
14       15   49      F       educator    97301
15       16   21      M  entertainment    10309
16       17   30      M     programmer    06355
17       18   35      F          other    37212
18       19   40      M      librarian    02138
19       20   42      F      homemaker    95660
20       21   26      M         writer    30068
21       22   25      M         writer    40206
22       23   30      F         artist    48197
23       24   21      F         artist    94533
24       25   39      M       engineer    55107

### Step 5. See the last 10 entries


####Option A

In [28]:
df_occupation.tail(10)

user_id  age gender     occupation zip_code
933      934   61      M       engineer    22902
934      935   42      M         doctor    66221
935      936   24      M          other    32789
936      937   48      M       educator    98072
937      938   38      F     technician    55038
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

####Option B

In [29]:
df_occupation.iloc[-10:]

user_id  age gender     occupation zip_code
933      934   61      M       engineer    22902
934      935   42      M         doctor    66221
935      936   24      M          other    32789
936      937   48      M       educator    98072
937      938   38      F     technician    55038
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

###Step 6. See the number of entries

####Option A

In [30]:
df_occupation.shape[0]

943

####Option B

In [31]:
df_occupation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user_id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      943 non-null    object
 3   occupation  943 non-null    object
 4   zip_code    943 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


###Step 7. See the number of columns

In [32]:
df_occupation.shape[1]

5

###Step 8. See all the columns' names.

####Option A

In [33]:
df_occupation.columns

Index(['user_id', 'age', 'gender', 'occupation', 'zip_code'], dtype='object')

####Option B

In [34]:
df_occupation.head(0)

Empty DataFrame
Columns: [user_id, age, gender, occupation, zip_code]
Index: []

 #### Option C
 concatenate the column names into a single string

In [35]:
print("The columns of the dataset are:")
print(", ".join(df_occupation.columns))

The columns of the dataset are:
user_id, age, gender, occupation, zip_code


###Step 9. See if the dataset is indexed. If yes, how?

In [36]:
df_occupation.index


RangeIndex(start=0, stop=943, step=1)

In fact, the dataset is indexed. The index ranges from 0 (start=0) to 943 (stop=943) in steps of one.

###Step 10. See the type of each column

DataFrame.dtypes returns a Series with the data type of each column

In [37]:
df_occupation.dtypes

user_id        int64
age            int64
gender        object
occupation    object
zip_code      object
dtype: object

###Step 11. See the information in the occupation column.

####Option A

In [38]:
df_occupation.occupation

0         technician
1              other
2             writer
3         technician
4              other
           ...      
938          student
939    administrator
940          student
941        librarian
942          student
Name: occupation, Length: 943, dtype: object

####Option B

In [39]:
df_occupation["occupation"]

0         technician
1              other
2             writer
3         technician
4              other
           ...      
938          student
939    administrator
940          student
941        librarian
942          student
Name: occupation, Length: 943, dtype: object

###Step 12. See how many distinct values exist in the occupation column?

####Option A

Answer in one line:

In [40]:
num_of_occupations1 = df_occupation.occupation.value_counts().count()
num_of_occupations1

21

Explanation:
The 5 first lines in the dataFrame are:

In [41]:
df_occupation.head()

user_id  age gender  occupation zip_code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

create a variable called occupation_column for saving only the occupation column data

In [42]:
occupation_column =  df_occupation.occupation
occupation_column

0         technician
1              other
2             writer
3         technician
4              other
           ...      
938          student
939    administrator
940          student
941        librarian
942          student
Name: occupation, Length: 943, dtype: object

to see how many 'students' we have in the column we can use .value_counts(). The resulting object will be in descending order so that the first element is the most frequently-occurring element. Excludes NA values by default.

In [43]:
num_val_count_occupations1 = occupation_column.value_counts()
num_val_count_occupations1

student          196
other            105
educator          95
administrator     79
engineer          67
programmer        66
librarian         51
writer            45
executive         32
scientist         31
artist            28
technician        27
marketing         26
entertainment     18
healthcare        16
retired           14
lawyer            12
salesman          12
none               9
homemaker          7
doctor             7
Name: occupation, dtype: int64

by using .count() we count non-NA cells for each column or row.

In [44]:
total_val_count_occupations1 = num_val_count_occupations1.count()
total_val_count_occupations1

21

Shape Return a tuple representing the dimensionality of the DataFrame and not for each colum/row.

In [45]:
num_val_count_occupations1.shape[0]

21

####Option B

The `nunique()` method returns the **number** of unique values for each column.

Answer in one line:

In [46]:
df_occupation.groupby('occupation').nunique().shape[0]

21

DataFrame.nunique(axis=0, dropna=True) Count number of distinct elements in specified axis ( index in this example). For example, there are 79 different administrator and 35 different ages. using shape will give as the number of different occupations becouse it is the same as the shape of the dataFrame. we should use

In [47]:
num_val_count_occupations2 = df_occupation.groupby('occupation').nunique()
num_val_count_occupations2

user_id  age  gender  zip_code
occupation                                   
administrator       79   35       2        79
artist              28   19       2        28
doctor               7    7       1         7
educator            95   37       2        91
engineer            67   31       2        64
entertainment       18   14       2        18
executive           32   21       2        32
healthcare          16   15       2        16
homemaker            7    7       2         7
lawyer              12   12       2        12
librarian           51   29       2        49
marketing           26   17       2        26
none                 9    9       2         9
other              105   39       2        99
programmer          66   29       2        63
retired             14    8       2        14
salesman            12   10       2        12
scientist           31   20       2        31
student            196   28       2       179
technician          27   16       2        27
writer              45   26       2        45

In [48]:
num_val_count_occupations2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21 entries, administrator to writer
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   user_id   21 non-null     int64
 1   age       21 non-null     int64
 2   gender    21 non-null     int64
 3   zip_code  21 non-null     int64
dtypes: int64(4)
memory usage: 840.0+ bytes


In [49]:
num_val_count_occupations2.shape[0]

21

####Option B

In [50]:
array = df_occupation.occupation.unique()
print(len(array))

21


###Step 13. See  which element is the most frequently-occurring element in the occupation column.

####Option A

 in step 12 option A we used .value_counts(). The resulting object will be in descending order so that the first element is the most frequently-occurring element. we saw that the answer is 'student'


```
num_of_occupations1 = df_occupation.occupation.value_counts().count()
```



In [51]:
num_val_count_occupations1

student          196
other            105
educator          95
administrator     79
engineer          67
programmer        66
librarian         51
writer            45
executive         32
scientist         31
artist            28
technician        27
marketing         26
entertainment     18
healthcare        16
retired           14
lawyer            12
salesman          12
none               9
homemaker          7
doctor             7
Name: occupation, dtype: int64

In [52]:
# num_val_count_occupations1 = df_occupation.occupation.value_counts()
num_val_count_occupations1.head(1)

student    196
Name: occupation, dtype: int64

####Option B

**better way : using the idxmax() method.**

this method returns a Series with the index of the maximum value for each column. By specifying the column axis ( axis='columns' ), the idxmax() method returns a Series with the index of the maximum value for each row.


In [53]:
df_occupation.occupation.value_counts()

student          196
other            105
educator          95
administrator     79
engineer          67
programmer        66
librarian         51
writer            45
executive         32
scientist         31
artist            28
technician        27
marketing         26
entertainment     18
healthcare        16
retired           14
lawyer            12
salesman          12
none               9
homemaker          7
doctor             7
Name: occupation, dtype: int64

In [54]:
df_occupation.occupation.value_counts().idxmax()

'student'

####Option C

Another option is using DataFrame.mode. It Get the mode(s) of each element along the selected axis.

The mode of a set of values is **the value that appears most often**. It can be multiple values.

In [55]:
df_occupation.occupation.mode()

0    student
Name: occupation, dtype: object

###Step 14. See the age column's descriptive statistics.
descriptive statistics; The number of columns, the average, the standard deviation, the percentiles of 25, 50, 75, the minimum and maximum

using the describe() method.this method generate descriptive statistics.

Descriptive statistics include those that summarize the central tendency, dispersion and shape of a dataset’s distribution, excluding NaN values.

In [56]:
age_column = df_occupation.age
age_column

0      24
1      53
2      23
3      24
4      33
       ..
938    26
939    32
940    20
941    48
942    22
Name: age, Length: 943, dtype: int64

In [57]:
age_column.describe()

count    943.000000
mean      34.051962
std       12.192740
min        7.000000
25%       25.000000
50%       31.000000
75%       43.000000
max       73.000000
Name: age, dtype: float64

###Step 15. See all column's descriptive statistics

In [58]:
df_occupation.describe(include = 'all')

user_id         age gender occupation zip_code
count   943.000000  943.000000    943        943      943
unique         NaN         NaN      2         21      795
top            NaN         NaN      M    student    55414
freq           NaN         NaN    670        196        9
mean    472.000000   34.051962    NaN        NaN      NaN
std     272.364951   12.192740    NaN        NaN      NaN
min       1.000000    7.000000    NaN        NaN      NaN
25%     236.500000   25.000000    NaN        NaN      NaN
50%     472.000000   31.000000    NaN        NaN      NaN
75%     707.500000   43.000000    NaN        NaN      NaN
max     943.000000   73.000000    NaN        NaN      NaN

###Step 16. See the most common occupation

we showed in step 13 the same answer

In [59]:
df_occupation.occupation.value_counts().idxmax()

'student'

In [60]:
df_occupation.occupation.mode()

0    student
Name: occupation, dtype: object